src/custom-element.js

```diff
@@ -123,48 +123,55 @@ export function stubElements(win) {
  * @param {!AmpElement} element
  */
 export function applyLayout_(element) {
+  const layoutAttr = element.getAttribute('layout');
   let widthAttr = element.getAttribute('width');
   let heightAttr = element.getAttribute('height');
   const sizesAttr = element.getAttribute('sizes');
-  const layoutAttr = element.getAttribute('layout');
+  let layout = null;
+  let width = null;
+  let height = null;
+
+  // Layout specified directly.
+  if (layoutAttr) {
+    layout = parseLayout(layoutAttr.trim());
+    if (!layout) {
+      throw new Error('Unknown layout: ' + layoutAttr);
+    }
+  }
 
-  // Handle elements that do not specify a width/height and are defined to have
-  // natural browser dimensions.
-  if ((!layoutAttr || layoutAttr == Layout.FIXED ||
-          layoutAttr == Layout.FIXED_HEIGHT) &&
+  // Default width and height: handle elements that do not specify a
+  // width/height and are defined to have natural browser dimensions.
+  if ((!layout || layout == Layout.FIXED || layout == Layout.FIXED_HEIGHT) &&
       (!widthAttr || !heightAttr) && hasNaturalDimensions(element.tagName)) {
     const dimensions = getNaturalDimensions(element.tagName);
-    if (layoutAttr != Layout.FIXED_HEIGHT) {
+    if (layout != Layout.FIXED_HEIGHT) {
       widthAttr = widthAttr || dimensions.width;
     }
     heightAttr = heightAttr || dimensions.height;
   }
 
-  let layout;
-  if (layoutAttr) {
-    // TODO(dvoytenko): show error state visually in the dev mode, e.g.
-    // red background + error message.
-    layout = parseLayout(layoutAttr.trim());
-    if (!layout) {
-      throw new Error('Unknown layout: ' + layoutAttr);
-    }
-  } else if (widthAttr || heightAttr) {
-    if (!widthAttr || widthAttr == 'auto') {
-      layout = Layout.FIXED_HEIGHT;
+  // Default layout value.
+  if (!layout) {
+    if (widthAttr || heightAttr) {
+      if (!widthAttr || widthAttr == 'auto') {
+        layout = Layout.FIXED_HEIGHT;
```

Comment:
If amp wants to default width and height first, that's ok. But then it's not allowed to use layout in doing so - see line 144, where it's checking layout to particular values. That's the circularity I mean here...
Note that the circularity gives rise to rather funky interpretations based on the data returned by getNaturalDimensions. For instance, assume it returns {width: 'auto', height: '1px'} (yes I know it doesn't), and the given width/height attributes are not set. Now it'll affect l. 156 and set layout to FIXED_HEIGHT. Resolving the circularity leads to a stricter behavior, which is desirable in my opinion, because it's easier to understand locally. Specifically, the choice of layout no longer depends on the return value of getNaturalDimensions - it's sufficient to know that an element hasNaturalDimensions, and the actual dimensions are only used to set width and height.